### ISSUE TITLE
Browsable API can't resolve URLs for HyperlinkedRelatedFields when using namespaced versioning

### ISSUE THREAD

**Issue by brandoncazander** [tomchristie/django-rest-framework#2489]
My POST requests to a ModelViewSet using a HyperlinkedModelSerializer don't work due to `HyperlinkedRelatedField.to_internal_value()` failing with 

```python
ValidationError: [u'Invalid hyperlink - Incorrect URL match.']
```

This is using NamespaceVersioning, and the failure originates at [line 225](https://github.com/tomchristie/django-rest-framework/blob/version-3.1/rest_framework/relations.py#L225) of relations.py since `match.view_name = 'v1:snippet-detail'` but `self.view_name = 'snippet-detail'`.

I have a failing unit test for this but can't decide how to best solve it. When we reverse URLs, if a versioning scheme is in place then the scheme's reverse function handles it (in particular, `NamespaceVersioning.reverse()` adds on the version to the view name).

Is there a way to get at the request from the HyperlinkedRelatedField? If I can get pointed in the right direction then I'll make up a pull request for this.

Thanks!

**Comment by tomchristie** [comment]
> Is there a way to get at the request from the HyperlinkedRelatedField?

Yes. https://github.com/tomchristie/django-rest-framework/blob/version-3.1/rest_framework/relations.py#L234

Wondering if we need to provide a `rest_framework.resolve` which is similar to `rest_framework.reverse` - takes a path and a request returns the resolver match, hooking into the versioning scheme on the way.

**Comment by brandoncazander** [comment]
I've done some work in the referenced commit on a possible `rest_framework.resolve`.

It would also be possible to handle this in `HyperlinkedRelatedField.to_internal_value`, by comparing the view_names only without the matched namespace.

**Comment by tomchristie** [comment]
@brandoncazander That looks good - would you mind issuing it as a pull request? (One inline comment worth resolving along the way.)

**Comment by brandoncazander** [comment]
@tomchristie Sorry for the wait! The Django 1.4 tests were failing and it took me a while to see why.

**Comment by tomchristie** [comment]
Actually I think the approach I was suggesting was wrong. Perhaps it's not `resolve` that we need, but actually a `get_versioned_viewname(viewname, request)`, like so:

    def get_versioned_viewname(viewname, request=None):
        """
        Hooks into the request versioning scheme to determine the versioned variant of a given view name.

        Normally this will simply return the viewname as-is, but in the case of namespaced versioning,
        the version prefix will be prefixed. If request is `None` the viewname will always be returned as is.
        """
        ...

We'd then use that in the URL check.

    viewname = get_versioned_viewname(self.view_name, request)
    if match.view_name != view_name:

Not sure the `resolve` approach makes sense?
Does this re-thinking seem sane to you, or am i just going round in circles?

**Comment by tomchristie** [comment]
Urg, unsure. Maybe the resolve approach is okay - it's really is the proper complement to the `reverse` function that we include, so...

Hrm.

**Comment by tomchristie** [comment]
Closed via #2505.
